napari/utils/colormaps/colormap.py

```diff
@@ -460,48 +580,205 @@ def _zero_preserving_modulo_numpy(
-                result.flat[i] = (values.flat[i] - 1) % n + 1
+    return result
+
+
+def _cast_labels_data_to_texture_dtype_direct(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    data = _convert_small_ints_to_unsigned(data)
+
+    if data.itemsize <= 2:
+        return data
 
-        return result
+    return _cast_labels_data_to_texture_dtype_direct_impl(
+        data, direct_colormap
+    )
+
+
+def _cast_labels_data_to_texture_dtype_direct_numpy(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """
+    Cast direct labels to the minimum type.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        The input data array.
+    direct_colormap : DirectLabelColormap
+        The direct colormap.
+
+    Returns
+    -------
+    np.ndarray
+        The casted data array.
+    """
+    max_value = max(x for x in direct_colormap.color_dict if x is not None)
+    if max_value > 2**16:
+        raise RuntimeError(  # pragma: no cover
+            "Cannot use numpy implementation for large values of labels "
+            "direct colormap. Please install numba."
+        )
+    dtype = minimum_dtype_for_labels(direct_colormap._unique_colors_num + 2)
+    label_mapping = direct_colormap._values_mapping_to_minimum_values_set()[0]
+
+    mapper = np.full((max_value + 2), DEFAULT_VALUE, dtype=dtype)
+    for key, val in label_mapping.items():
+        if key is None:
+            continue
+        mapper[key] = val
+
+    if data.dtype.itemsize > 2:
+        data = np.clip(data, 0, max_value + 1)
+    return mapper[data]
+
+
+def _generate_hash_map_for_direct_colormap(
+    direct_colormap: DirectLabelColormap,
+    data_dtype: np.dtype,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Generate hash map for direct colormap.
+    """
+    target_dtype = minimum_dtype_for_labels(
+        direct_colormap._unique_colors_num + 2
+    )
+    label_mapping = direct_colormap._values_mapping_to_minimum_values_set()[0]
+    prime_num_array = _primes(upto=2**16)
+    pos = bisect.bisect_left(prime_num_array, len(label_mapping) * 2)
+    if pos < len(prime_num_array):
+        hash_size = prime_num_array[pos]
+    else:
+        hash_size = 2 ** (math.ceil(math.log2(len(label_mapping))) + 1)
+
+    hash_table_key = np.zeros(hash_size, dtype=data_dtype)
+    hash_table_val = np.zeros(hash_size, dtype=target_dtype)
+
+    data_min = np.iinfo(data_dtype).min
+    data_max = np.iinfo(data_dtype).max
+
+    for key, val in label_mapping.items():
+        if key is None:
+            continue
+        if key > data_max or key < data_min:
+            continue
+        new_key = key % hash_size
+        while hash_table_key[new_key] != 0:
+            new_key = (new_key + 1) % hash_size
+
+        hash_table_key[new_key] = key
+        hash_table_val[new_key] = val
+
+    return hash_table_key, hash_table_val
+
+
+def _generate_hash_map_and_cast_labels_data_to_texture_dtype_direct(
```

Comment:
@Czaki I just realised that these are the same as [`skimage.util.map_array`](https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.map_array) (compute hash table and do translation together) and [`skimage.util._map_array.ArrayMap`](https://github.com/scikit-image/scikit-image/blob/441fe68b95a86d4ae2a351311a0c39a4232b6521/skimage/util/_map_array.py#L61-L105) (cached hash table). They were created for exactly this use case 😅 and @andy-sweet even investigated speeding them up for us! (Albeit unsuccessfully — we were looking at L1 cache-access changes, not parallelising.) Anyway, because it's not parallelised, it's about 10x slower than this Numba version on my machine. However (a) we can use it instead of the NumPy-based mapping when Numba is not installed — it doesn't suffer from the same size limitations, so we can get rid of the "please install Numba" message, and (b) if you want to have a look at speeding up that Cython in skimage, that would be a nice way to upstream this work. 😊 

I'm actually working on a bunch of naming changes for this PR, I can also work on changing NumPy to ArrayMap if you like?

Reply:
Hmmm, nope, [there is no caching of the hash table](https://github.com/scikit-image/scikit-image/blob/441fe68b95a86d4ae2a351311a0c39a4232b6521/skimage/util/_map_array.py#L172-L176), I guess because there's no easy way to refer to a C++ hash_map from Python. Still, this can replace the NumPy approach I think.

Reply:
Another note that I need to play around with — Numba accelerates typed dictionaries, so we might be better off using that rather than implementing our own dict with linear collision resolution...

Reply:
I spot typed dict yesterday evening. It looks interesting.

Reply:
@jni Did you know how `skimage.util.map_array` deal with missed values?

Reply:
map array deal good with missing values, but is slow:

here is a script that I have created (required ipython):
```python
import numpy as np

from napari.utils.colormaps.colormap import _labels_raw_to_texture_direct_loop, \
    _labels_raw_to_texture_direct_scikit_map, _labels_raw_to_texture_direct_typed_dict, DirectLabelColormap

data = np.random.randint(0, 100000, size=(2000, 2000), dtype=np.uint32)
info = np.iinfo(np.uint32)

random_label_ids = np.random.randint(
            low=max(-10000, info.min), high=min(10000, info.max), size=20
        )

color = {i + 1: np.random.random(4) for i in random_label_ids}
cmap = DirectLabelColormap(np.array([0,0,0]), color=color)

m1 = _labels_raw_to_texture_direct_loop(data, cmap)
m2 = _labels_raw_to_texture_direct_scikit_map(data, cmap)
m3 = _labels_raw_to_texture_direct_typed_dict(data, cmap)

assert np.allclose(m1, m2)
assert np.allclose(m1, m3)

%timeit _labels_raw_to_texture_direct_loop(data, cmap)
%timeit _labels_raw_to_texture_direct_scikit_map(data, cmap)
%timeit _labels_raw_to_texture_direct_typed_dict(data, cmap)
```

Results:
```
1 ms ± 18.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
18 ms ± 271 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.33 ms ± 110 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

with implementation:
```python
def _labels_raw_to_texture_direct_scikit_map(
    data: np.ndarray, direct_colormap: DirectLabelColormap
):
    return map_array(data, * direct_colormap._mapping_as_arrays)

def _labels_raw_to_texture_direct_typed_dict(
    data: np.ndarray, direct_colormap: DirectLabelColormap
):
    return _labels_raw_to_texture_direct_typed_dict_impl(
        data, direct_colormap._typed_dict_mapping
    )

import numba

@numba.njit(parallel=True)
def _labels_raw_to_texture_direct_typed_dict_impl(
    data: np.ndarray, mapping: numba.typed.Dict
):
    result = np.empty_like(data, dtype=np.uint8)
    for i in prange(data.size):
        val = data.flat[i]
        if val in mapping:
            result.flat[i] = mapping[val]
    return result
```

and new cached properties:

```python
    @cached_property
    def _mapping_as_arrays(self):
        mapping = self._values_mapping_to_minimum_values_set()[0]
        res1, res2 = [], []
        for key, value in mapping.items():
            if key is None:
                continue
            res1.append(key)
            res2.append(value)
        return np.array(res1), np.array(res2)

    @cached_property
    def _typed_dict_mapping(self):
        from numba.typed import Dict
        from numba.core import types
        mapping = self._values_mapping_to_minimum_values_set()[0]
        res = Dict.empty(
            key_type=types.uint32,
            value_type=types.uint8,
        )
        res.update({key: value for key, value in mapping.items() if key is not None})
        return res
```

Reply:
Cool, @Czaki can you also add the NumPy implementation to the comparison?

Reply:
There was a stupid bug in the code (I pas color instead of color_dict to DirectColormap constructor):

```
import numpy as np

from napari.utils.colormaps.colormap import _labels_raw_to_texture_direct_loop, \
    _labels_raw_to_texture_direct_scikit_map, _labels_raw_to_texture_direct_typed_dict, DirectLabelColormap, _labels_raw_to_texture_direct_numpy

data = np.random.randint(0, 100000, size=(2000, 2000), dtype=np.uint32)
info = np.iinfo(np.uint32)

random_label_ids = np.random.randint(
            low=max(-10000, info.min), high=min(10000, info.max), size=200
        )

color = {i + 1: np.random.random(4) for i in random_label_ids}
cmap = DirectLabelColormap(np.array([0,0,0]), color_dict=color)

m1 = _labels_raw_to_texture_direct_loop(data, cmap)
m2 = _labels_raw_to_texture_direct_scikit_map(data, cmap)
m3 = _labels_raw_to_texture_direct_typed_dict(data, cmap)
m4 = _labels_raw_to_texture_direct_numpy(data, cmap)

assert np.allclose(m1, m2)
assert np.allclose(m1, m3)
assert np.allclose(m1, m4)

%timeit _labels_raw_to_texture_direct_loop(data, cmap)
%timeit _labels_raw_to_texture_direct_scikit_map(data, cmap)
%timeit _labels_raw_to_texture_direct_typed_dict(data, cmap)
%timeit _labels_raw_to_texture_direct_numpy(data, cmap)
```

results:

```
10.3 ms ± 368 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17.1 ms ± 71 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.73 ms ± 98.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.75 ms ± 26.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```